### PR TITLE
Preload Assets

### PR DESCRIPTION
**Pull request by goodguyry**
This adds `am_preload` for preloading assets of any ([supported](https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content#What_types_of_content_can_be_preloaded)) type.

```php
// `as` and `mime_type` options will be automatically added for CSS, but are included here for clarity.
am_preload(
	[
		'handle'    => 'preload-styles',
		'src'       => 'css/styles.css',
		'condition' => 'global',
		'version'   => '1.0.0',
		'as'        => 'style'
		'mime_type' => 'text/css',
	]
);
```

```html
<link rel="preload" href="http://client/css/test.css?ver=1.0.0" class="wp-asset-manager preload-styles" as="style" media="all" type="text/css" />
```

The `am_preload` function patches the `as` and `mime_type` option values for common use-cases (CSS, JavaScript and WOFF2 fonts), but will throw an error if the `as` option is missing for any other file type.

From [the spec](https://www.w3.org/TR/preload/#as-attribute):
> The [as] attribute is necessary to guarantee correct prioritization, request matching, application of the correct Content Security Policy policy, and setting of the appropriate `Accept` request header.

This function will also automatically add the `crossorigin` attribute for fonts, which [is required](https://drafts.csswg.org/css-fonts/#font-fetching-requirements) when preloading fonts, even if they're not actually cross-origin requests.

More on preloading here: https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content

### CODE REVIEW

File: php/class-asset-manager-preload.php [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	 * Constructor.
+	 */
+	private function __construct() {
+		// Don't do anything, needs to be initialized via instance() method.
+	}
+
+	/**
+	 * Get an instance of the class.
+	 *
+	 * @return class
```
Comment (mslinnea) [inline php/class-asset-manager-preload.php]: This phpdoc should be `@return Asset_Manager_Preload` ... An object is being returned and that kind specifically :)

File: php/class-asset-manager-preload.php [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	 * @return void
+	 */
+	public function print_asset( $asset ) {
+		$classes      = $this->default_classes;
+		$classes[]    = $asset['handle'];
+		$print_string = '';
+
+		if ( empty( $asset['as'] ) || ! in_array( $asset['as'], $this->preload_as, true ) ) {
+			// We weren't able to patch in the 'as' attribute in `post_validate_asset`.
+			$this->generate_asset_error( 'invalid_preload_attribute', $asset, [ 'attribute' => 'as' ] );
```
Comment (mslinnea) [inline php/class-asset-manager-preload.php]: How about adding this error code to the function generate_asset_error so that a specific error is displayed?

**Review by mslinnea — DISMISSED** [review]
This is looking real good!! 🐴

File: php/class-asset-manager-preload.php (hunk 2 above)
Comment (goodguyry) [reply php/class-asset-manager-preload.php]: Hmm... I thought I did. Thanks! I'll double-check to make sure.

**Review by mslinnea — APPROVED** [review]
🐴

**Pull request merged**